ch6/boggle.py

```diff
@@ -0,0 +1,59 @@
+from collections import defaultdict
+
+SIZE_OF_BOARD = 5
+MAX_NUM_OF_ALPHABET_PER_WORD = 10
+dx = [-1, -1, -1, 1, 1, 1, 0, 0]
+dy = [-1, 0, 1, -1, 0, 1, -1, 1]
+
+def in_range(y, x): 
+    return x >= 0 and y >= 0 and x <= 4 and y <= 4
+
+
+def has_word(y, x, word, cur_index, cache):
+    if not in_range(y, x):
+        return False
+    cache[y][x][cur_index] = 1
+    if board[y][x] != word[cur_index]:
+        return False
+    if cur_index == len(word)-1:
+        return True
+    for direction in xrange(8):
+        nextY = y + dy[direction]
+        nextX = x + dx[direction]
+        if not in_range(nextY, nextX):
+            continue
+        if cache[nextY][nextX][cur_index+1]:
+            continue
+        if has_word(nextY, nextX, word, cur_index+1, cache):
+            return True
+    return False
+
+
+def initialize_cache():
```

Comment:
I think a more pythonic solution is to use a 3-tuple.

e.g.

cache[(y, x, cur_index)] = ...

instead of

cache[y][x][cur_index]

In a language like C, allocating a fixed size data structure is more natural.
in a language like Python, allocating a dynamically growing data structure is more natural.

The omitted part of the C(++) solution most likely initializes a cache of some fixed size
e.g.
int cache[UPPER_BOUND_ON_Y][UPPER_BOUND_ON_X][UPPER_BOUND_ON_INDEX];

where UPPER_BOUND_ON_XXX is the maximum possible of the corresponding dimension.